### ISSUE TITLE
[ no-constant-condition] should validate 'void' too 

### ISSUE THREAD

**Issue by pvamshi** [eslint/eslint#5726]
**Request type** : Enhancement 

**Rule** : `no-constant-condition`

**What version of ESLint are you using?**
2.5.1

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```js
{
  "rules": {
    "no-constant-condition": 2   // disallow use of constant expressions in conditions
  }
}
```
**What did you do? Please include the actual source code causing the issue.**
```js
if(void x){}
if (void x === void x){}
````
**What did you expect to happen?**
Error as it is a constant condition

**What actually happened?**
No error

As discussed in #5228 &  #5696

**Comment by ilyavolodin** [comment]
Sounds reasonable to me. Although I think the primary case should be:
```js
if (void) {}
```

**Comment by isiahmeadows** [comment]
@ilyavolodin You mean `if (void foo) {}`? :wink:

**Comment by ilyavolodin** [comment]
Yes I do:-)

**Comment by vitorbal** [comment]
I would love to tackle this one if it gets accepted!

**Comment by michaelficarra** [comment]
There's also some other missing expression types in the `isConstant` function. ConditionalExpression stands out to me. I think this issue should be about generally expanding that function as best we can.

**Comment by vitorbal** [comment]
Should it really be addressed in this issue? I have a PR ready for the `void` case otherwise.

**Comment by alberto** [comment]
Is anyone championing this? @eslint/eslint-team

**Comment by platinumazure** [comment]
I count one :+1: (@ilyavolodin, maybe a stretch). I will add one more :+1:.

@vitorbal, you could champion or add :+1:. Then we need whatever you don't do. :smile:

**Comment by vitorbal** [comment]
@alberto @platinumazure thanks for the ping! I am willing to champion this change and I've already got a PR almost ready, just in need of a rebase.

I agree that we should open a separate issue for the extra checks.

**Comment by nzakas** [comment]
I'll 👍  

@vitorbal as the champion, you should secure one more definitive 👍 and then you can mark as accepted.

**Comment by alberto** [comment]
👍

**Comment by vitorbal** [comment]
Thanks all, I will get around to finishing this by the end of this week!
